### PR TITLE
Sketch-Guided Equality Saturation

### PR DESCRIPTION
**Pull request by Bastacyclop**
Uses https://github.com/elevate-lang/elevate/pull/28
comments:
- the `RewriteDirected` mechanism improves on previous attempts, but is still bugged (problem with e-class ids and find)
- beam extraction is implemented but lacks a way to promote diversity inside the beam
- memory is pretty printed in e.g. kibibytes, but maybe kilobytes are more familiar to read and compute with
- the `separableConvolution2DRewrite` test is now counting rule applications in one case, defining a special `traversableCountApplications` which is very redundant with other Rise traversal in Elevate.

### CODE REVIEW

File: src/main/scala/rise/eqsat/ExtendedPattern.scala [hunk 1]
```diff
@@ -0,0 +1,753 @@
...
+    assert(egraph.clean)
+    searchEClass(egraph, egraph.get(id), Set(), HashMap.empty/*, None*/)
+  }
+
+  // TODO: investigate similarities with e-matching
+  private def searchEClass(egraph: EGraph,
+                           eclass: EClass,
+                           visited: Set[EClassId],
+                           memo: HashMap[(EClassId, ExtendedPattern), Vec[ExtendedPatternMatch]],
+                           /*parent: Option[ENode]*/): Vec[ExtendedPatternMatch] = {
```
Comment (Bastacyclop) [inline src/main/scala/rise/eqsat/ExtendedPattern.scala:520]: This search method is technically not used anymore .. should we keep the code around or delete it?

File: src/main/scala/rise/eqsat/Rewrite.scala [hunk 2]
```diff
@@ -3,7 +3,7 @@ package rise.eqsat
 import rise.core.{types => rct}
 
 object Rewrite {
-  def init[ED, ND, DT](name: String, rule: (Searcher[ED, ND, DT], Applier[ED, ND, DT])): Rewrite[ED, ND, DT] = {
+  def init[ED, ND, DT](name: String, rule: (Searcher, Applier)): Rewrite = {
```
Comment (michel-steuwer) [inline src/main/scala/rise/eqsat/Rewrite.scala:6]: Are these type parameters (ED, ND, DT) still required?

**Review by michel-steuwer — APPROVED** [review]
Only some minor comments that I had while browsing the code. Obviously, I haven't looked at all code detailedly.

File: src/main/scala/rise/eqsat/Predicate.scala [hunk 3]
```diff
@@ -1,33 +1,39 @@
 package rise.eqsat
 
+import arithexpr.arithmetic.BoolExpr.ArithPredicate
 import rise.eqsat.ematching.MNode
 
 // TODO: think about what the predicate interface should be
```
[the remark this answers is not in the document]
Comment (Bastacyclop) [reply src/main/scala/rise/eqsat/Predicate.scala:6]: The predicate trait remains ad-hoc without much generic design effort, I can remove the comment.

File: src/main/scala/rise/eqsat/Node.scala [hunk 4]
```diff
@@ -8,10 +8,11 @@ import rise.core.semantics
   * @tparam N abstracts over contained nats
   * @tparam DT abstracts over contained data types
   */
-sealed trait Node[+E, +N, +DT] {
-  def map[OE, ON, ODT](fe: E => OE,
-                       fn: N => ON,
-                       fdt: DT => ODT): Node[OE, ON, ODT] = this match {
+sealed trait Node[+E, +N, +DT, +A] {
```
[the remark this answers is not in the document]
Comment (Bastacyclop) [reply src/main/scala/rise/eqsat/Node.scala:11]: :+1:

File: src/main/scala/rise/eqsat/Rewrite.scala (hunk 2 above)
Comment (Bastacyclop) [reply src/main/scala/rise/eqsat/Rewrite.scala:6]: No, they are dead code!

**Comment by Bastacyclop** [comment]
Thanks for the review @michel-steuwer , I have adressed your comments.

**Review by michel-steuwer — APPROVED** [review]

**Pull request merged**